Fix missed places in dataflow analysis that introduce genericness

MichalStrehovsky opened the pull request:
Resolves #117998.

Fixes the cases in the bug and adds a couple more obscure corner cases.

Cc @dotnet/ilc-contrib

MichalStrehovsky committed:
Fix

MichalStrehovsky committed:
Test

sbomer reviewed (approved):
Am I right in thinking that these are the same cases we would need to detect for https://github.com/dotnet/runtime/issues/90115?

MichalStrehovsky committed:
ILLinker warns more on this?

Unexpected warning found: ILLink: Trim analysis warning IL2091: Mono.Linker.Tests.Cases.DataFlow.GenericParameterDataFlow.TestInArray<T>(): 'T' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in 'Mono.Linker.Tests.Cases.DataFlow.GenericParameterDataFlow.RequiresParameterlessCtor<T>'. The generic parameter 'T' of 'Mono.Linker.Tests.Cases.DataFlow.GenericParameterDataFlow.TestInArray<T>()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
HOW MANY MORE?

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs

MichalStrehovsky committed:
Update GenericParameterDataFlow.cs
